massa-pos-exports/src/deferred_credits.rs

```diff
@@ -0,0 +1,192 @@
+                // address
+                buffer.extend(addr.to_bytes());
+                // credited amount
+                self.amount_ser.serialize(amount, buffer)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Deserializer for `DeferredCredits`
+pub struct DeferredCreditsDeserializer {
+    u64_deserializer: U64VarIntDeserializer,
+    slot_deserializer: SlotDeserializer,
+    credit_deserializer: CreditDeserializer,
+}
+
+impl DeferredCreditsDeserializer {
+    /// Creates a new `DeferredCredits` deserializer
+    pub fn new(thread_count: u8) -> DeferredCreditsDeserializer {
+        DeferredCreditsDeserializer {
+            u64_deserializer: U64VarIntDeserializer::new(Included(u64::MIN), Included(u64::MAX)),
+            slot_deserializer: SlotDeserializer::new(
+                (Included(0), Included(u64::MAX)),
+                (Included(0), Excluded(thread_count)),
+            ),
+            credit_deserializer: CreditDeserializer::new(),
+        }
+    }
+}
+
+impl Deserializer<DeferredCredits> for DeferredCreditsDeserializer {
+    fn deserialize<'a, E: ParseError<&'a [u8]> + ContextError<&'a [u8]>>(
+        &self,
+        buffer: &'a [u8],
+    ) -> IResult<&'a [u8], DeferredCredits, E> {
+        context(
+            "Failed DeferredCredits deserialization",
+            length_count(
+                context("Failed length deserialization", |input| {
+                    self.u64_deserializer.deserialize(input)
+                }),
+                tuple((
+                    |input| self.slot_deserializer.deserialize(input),
+                    |input| self.credit_deserializer.deserialize(input),
+                )),
+            ),
+        )
+        .map(|elements| DeferredCredits(elements.into_iter().collect()))
+        .parse(buffer)
+    }
+}
+
+/// Deserializer for a single credit
+struct CreditDeserializer {
+    u64_deserializer: U64VarIntDeserializer,
+    address_deserializer: AddressDeserializer,
+    amount_deserializer: AmountDeserializer,
+}
+
+impl CreditDeserializer {
+    /// Creates a new single credit deserializer
+    fn new() -> CreditDeserializer {
+        CreditDeserializer {
+            u64_deserializer: U64VarIntDeserializer::new(Included(u64::MIN), Included(u64::MAX)),
+            address_deserializer: AddressDeserializer::new(),
+            amount_deserializer: AmountDeserializer::new(
+                Included(Amount::MIN),
+                Included(Amount::MAX),
+            ),
+        }
+    }
+}
+
+impl Deserializer<PreHashMap<Address, Amount>> for CreditDeserializer {
+    fn deserialize<'a, E: ParseError<&'a [u8]> + ContextError<&'a [u8]>>(
+        &self,
+        buffer: &'a [u8],
+    ) -> IResult<&'a [u8], PreHashMap<Address, Amount>, E> {
+        context(
+            "Failed Credit deserialization",
+            length_count(
+                context("Failed length deserialization", |input| {
+                    self.u64_deserializer.deserialize(input)
```

Comment:
A maximum value value should be set for this deserializer to avoid allocating big vector